pkg/metrics/metrics.go

```diff
@@ -154,22 +156,65 @@ func (c *textFileCollector) Collect(ch chan<- prometheus.Metric) {
 
 // ProcessMetricsFile processes a metrics file that follows Prometheus text format: https://prometheus.io/docs/instrumenting/exposition_formats/,
 // returning its MetricFamily.
-func ProcessMetricsFile(path string) (map[string]*dto.MetricFamily, error) {
-	f, err := os.Open(path)
+func ProcessMetricsFile(directoryPath string) (map[string]*dto.MetricFamily, error) {
+	// Find latest file generation.
+	generationFilePath := directoryPath + generationFileName
+	genNumber, err := getGenerationNumber(generationFilePath)
 	if err != nil {
-		return nil, fmt.Errorf("failed to open metrics file %q: %w", path, err)
+		return nil, fmt.Errorf("could not get generation number from %s: %w", generationFilePath, err)
 	}
-	defer f.Close()
+
+	// Find latest metrics file name and open.
+	metricsFilePath := directoryPath + GetMetricsFileName(genNumber)
+	metricsFile, err := os.Open(metricsFilePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open metrics file %q: %w", metricsFilePath, err)
+	}
+	defer metricsFile.Close()
 
 	var parser expfmt.TextParser
-	metricFamilies, err := parser.TextToMetricFamilies(f)
+	metricFamilies, err := parser.TextToMetricFamilies(metricsFile)
 	if err != nil {
-		return nil, fmt.Errorf("failed to parse metrics file %q: %w", path, err)
+		return nil, fmt.Errorf("failed to parse metrics file %q: %w", metricsFilePath, err)
 	}
 
 	return metricFamilies, nil
 }
 
+// GetMetricsFilePath creates the expected name for the latest metrics file.
+func GetMetricsFileName(genNumber int) string {
+	return fmt.Sprintf(metricsFileNameTemplate, genNumber)
+}
+
+func GetGenerationFileName() string {
+	return generationFileName
+}
+
+// getGenerationNumber opens the generation.txt file and parses the payload into
+// an integer. This integer represents the current generation of the metrics file.
+func getGenerationNumber(filePath string) (int, error) {
```

Comment:
Is reading this file also prone to racing?

Reply:
The main concern with the original approach is that we were reading and writing large number of contents to our files which could cause the writer to write in chunks, and reader to see incomplete data. Here we are only writing a piece of information and closing. In this case, i believe its difficult to read corrupted data in this file.

Reply:
Possibly? It may still be worth using flock just to avoid the possibility.

Reply:
In that case would we still want to keep this implementation? 

The reason to steer away from file locks was that to properly apply them, we would have to lock on both ends (when reading and writing). Performing many reads on the endpoint could make us run into issues where we cant write because of lock. I will check if there's timing for how often we read this endpoint from metrics sidecar.

Reply:
See #310 for latest change.